### ISSUE TITLE
`yarn upgrade` refuses to work inside a workspace.

### ISSUE THREAD

**Issue by SamirTalwar** [yarnpkg/yarn#4442]
**Do you want to request a *feature* or report a *bug*?**

This is a bug.

**What is the current behavior?**

Inside a workspace, `yarn upgrade` fails immediately.

**If the current behavior is a bug, please provide the steps to reproduce.**

Given the following structure:

```
stuff $ tree
.
├── node_modules
│   ├── stuff-bar -> ../packages/bar
│   └── stuff-foo -> ../packages/foo
├── package.json
├── packages
│   ├── bar
│   │   └── package.json
│   └── foo
│       └── package.json
└── yarn.lock

6 directories, 4 files
```

Running `yarn upgrade` does this:

```
stuff/packages/foo $ yarn upgrade
yarn upgrade v1.0.2
error No lockfile in this directory. Run `yarn install` to generate one.
```

**What is the expected behavior?**

Yarn upgrades the packages referenced in the *package.json* file.

**Please mention your node.js, yarn and operating system version.**

node.js: 8.4.0 from Nix
Yarn: 1.0.2
OS: macOS 10.12.6

**Comment by edmorley** [comment]
Hi! I think the fixes in #4654 might have solved this. Could you give yarn 1.2.0 a go and see if it still reproduces? :-)

**Comment by mattnathan** [comment]
I have a similar issue but with the `--latest` option, yarn says it's doing what you expect but none of the dependencies or `package.json` are upgraded when within a workspace.

```
>yarn versions
yarn versions v1.3.2
{ http_parser: '2.7.0',
  node: '8.9.1',
  v8: '6.1.534.47',
  uv: '1.15.0',
  zlib: '1.2.11',
  ares: '1.10.1-DEV',
  modules: '57',
  nghttp2: '1.25.0',
  openssl: '1.0.2m',
  icu: '59.1',
  unicode: '9.0',
  cldr: '31.0.1',
  tz: '2017b' }
Done in 0.07s.
```

**Comment by pleunv** [comment]
I'm witnessing the same behavior with `--latest`. Only way around it seems to be doing a `yarn add package@<latest-version>`.

**Comment by EECOLOR** [comment]
It seems `yarn upgrade-interactive --latest` does work

**Comment by SamirTalwar** [comment]
Seems to work well now.

**Comment by EECOLOR** [comment]
@SamirTalwar For me it does not work with workspaces, I am on version 1.5.1 (latest). Another problem is that this part of [the documentation](https://yarnpkg.com/lang/en/docs/cli/upgrade/#toc-yarn-upgrade-package-latest-l-caret-tilde-exact-pattern) is not respected:

> The `package.json` file will be updated to reflect the latest version range.

**Comment by SamirTalwar** [comment]
Happy for this to be reopened until someone can definitely answer this one. I can’t reproduce it any more, but I haven’t tried very hard. I’m writing very little JavaScript at the moment.

**Comment by alxmiron** [comment]
I'm also suffering from the issue. Steps I can reproduce:

`yarn outdated` - see outdated packages, let's take `bugsnag-js` as an example
`yarn upgrade bugsnag-js --latest` - successful ugrade with result:
```
success Saved lockfile.
success Saved 7 new dependencies.
info Direct dependencies
info All dependencies
├─ @hitask/catalog@0.1.47
├─ @hitask/desktop@1.1.1
├─ @hitask/ext-app@3.1.1
├─ @hitask/ext-calendar@1.1.1
├─ @hitask/mock-api@0.1.9
├─ @hitask/resources@0.1.2
└─ @hitask/webapp@1.1.3
```

`yarn outdated` - no changes in the list. No changes in any package.json and yarn.lock too

**Comment by alxmiron** [comment]
`yarn upgrade-interactive --latest` is working ok

**Comment by pronebird** [comment]
Can't get this to work. The last comment since April 10... is it dead project?

```
yarn workspace %WORKSPACE% upgrade @babel/cli @babel/core babel-preset-react-native --latest
```

**Comment by indirectlylit** [comment]
Reproducing this issue on macos 10.14, yarn 1.15.2, and can confirm the `upgrade-interactive` workaround above.

**Comment by neodon** [comment]
This bug is still around in yarn v1.16.0.  Here is a minimal example repo: https://github.com/neodon/yarn-workspace-upgrade-bug

The repo has two workspaces, in `packages/a` and `packages/b`.  Both workspaces and the root have the `chai` module installed with the version set to exactly `4.0.0`.  The latest is `4.2.0`.

Running `yarn upgrade --latest` in the root does what I expect--the `chai` module is upgraded to `4.2.0` and both `package.json` and `yarn.lock` are updated.

root package.json diff:

```
   "dependencies": {
-    "chai": "4.0.0"
+    "chai": "4.2.0"
   }
```

Now I switch into the `packages/a` directory and run `yarn upgrade --latest`.  No changes.  Same if I run the command in `packages/b`.

I expect that when I run `yarn upgrade --latest` in a workspace directory, it operates on that workspace the same as the root or a non-workspace project.  It should both upgrade all packages to the latest version and update `package.json`.

If I run `yarn upgrade-interactive --latest` anywhere in the project, it lets me select the workspaces and it updates their `package.json` files as expected.  I think `yarn upgrade --latest` should do the same thing when run in a workspace directory.

**Comment by pldg** [comment]
I've tried `yarn upgrade-interactive --latest` in the root of the repository, it upgrades packages correctly in the root *node_modules*, but does not upgrade *packages.json* files inside `packages/*`.

**Comment by maciej-gurban** [comment]
Same as @pldg , there doesn't seem to be a way to upgrade a dependency of a workspace using `yarn upgrade`, or `yarn upgrade-interactive`. On yarn 1.16.0 and OSX 10.14.4 if that matters. Unlike for @neodon there's nothing that I'd describe as letting me "select workspaces". I only see the list of root level dependencies.

**Comment by o-alexandrov** [comment]
## Environment
MacOS version: `10.15.1`
Yarn installed via HomeBrew
Yarn version: `1.19.2`
## Usage
Yarn workspace
## Cases & Results
- [yarn upgrade-interactive --latest](https://yarnpkg.com/lang/en/docs/cli/upgrade-interactive/)
  - running this command from any of the directories in a workspace successfully finds all workspace packages outdated packages and prompts to upgrade them **interactively**
- [yarn upgrade --latest](https://yarnpkg.com/lang/en/docs/cli/upgrade/)
  - running this command from any directories, including from the package that has outdated deps results in **NO upgrades of the yarn.lock AND `package.json`**, moreover, this command **doesn't update the modules in the `node_modules/*`**
    - in short, doesn't work at all

**Comment by josh08h** [comment]
I've encountered the a similar issue, see https://github.com/yarnpkg/yarn/issues/7698 where I have created a reproducible repo.

Any update from the yarn team on this?

**Comment by sabdullahpear** [comment]
Encountering the same issue, `yarn upgrade --latest` doesn't touch the workspaces. Running the command from inside the workspace directory also doesn't work, so the only way to upgrade packages inside a workspace is by manually upgrading the packages one by one :(

**Comment by jazzfog** [comment]
Still not fixed.

**Comment by jazzfog** [comment]
I "solved" it in a different way, did not want to mess with interactive mode:
- Copied the `package.json` to an another dir and run there
- `yarn` first
- and then `yarn upgrade --latest`
- Moved the (updated by now) `package.json` back to the original dir and run there
- `yarn`

**Comment by parafox0417** [comment]
@jazzfog It appears that Yarn v1 is essentially abandoned since they are working on the Yarn v2 rewrite.

Yarn v2 is looking awesome, but their abandoning Yarn v1 really irks me and I'm on the fence about promoting Yarn v2 in my organization.

When will they decide it's time for Yarn v3 and abandon Yarn v2 the same way?  I think they need to address the community on this problem.

**Comment by jazzfog** [comment]
@parafox0417 You making a good point.

**Comment by guytt** [comment]
We are facing this issue as well. Very frustrating... Over 3 years now

**Comment by paul-soporan** [comment]
Closing as fixed in the v2 (use the `yarn up` command).

https://yarnpkg.com/getting-started/migration

**Comment by paul-soporan** [comment]
@parafox0417

> It appears that Yarn v1 is essentially abandoned since they are working on the Yarn v2 rewrite.

We're not only working on it, it's already been stable for a long time and the migration should now be pretty straight forward.

> Yarn v2 is looking awesome, but their abandoning Yarn v1 really irks me and I'm on the fence about promoting Yarn v2 in my organization.

We haven't abandoned Yarn 1, it's essentially been in implicit maintenance mode (aka nobody tackling issues like this) for quite some time before we made it explicit when we released Yarn 2. Yarn 1 will only receive security fixes and things that improve the migration process.

Also, it doesn't make sense for us to try to fix this in v1, as we've completely revamped the workspace support in v2 to prevent issues like this from even existing. We're not going to try finding workarounds for v1 when, by now, more than 99% of users should be able to upgrade to v2 with the `node-modules` linker without problems.

> When will they decide it's time for Yarn v3 and abandon Yarn v2 the same way? I think they need to address the community on this problem.

Yarn 2 is the last major rewrite, we now have a solid foundation to build upon in the following versions (we'll release Yarn 3 soon and it won't have many breaking changes).

**Comment by basickarl** [comment]
Any updates on this?

**Comment by dishui** [comment]
not working in workspace:
`yarn workspace %WORKSPACE% upgrade-interactive %PACKAGENAME% --latest `
give the error: `Usage Error: Couldn't find a script named "upgrade".`
and
`yarn workspace %WORKSPACE% upgrade %PACKAGENAME% --latest `
gives the error:
Unknown Syntax Error: Extraneous positional argument ("%PACAKGENAME").

both are not working.
